### PR TITLE
Clear the regenerator cache every time we process

### PR DESCRIPTION
**Pull request by nickburlett**
To address part of #3591, clear the regenerator's cache every time the
site is processed. This ensures that the regenerator doesn't incorrectly
believe a file hasn't changed based on stale information.

### CODE REVIEW

**Comment by nickburlett** [comment]
I'm not entirely sure how to deal with the other part of #3591 (where it doesn't cache or update metadata for all dependencies), but this at least fixes the bigger problem of not regenerating files that do need regeneration.

File: lib/jekyll/regenerator.rb [hunk 1]
```diff
@@ -59,6 +59,14 @@ def clear
       @cache = {}
```
Comment (parkr) [inline lib/jekyll/regenerator.rb]: if we go with this, this should just use that helper

File: lib/jekyll/site.rb [hunk 2]
```diff
@@ -68,6 +68,7 @@ def reset
       self.static_files = []
       self.data = {}
       @collections = nil
+      @regenerator.clear_cache()
```
Comment (parkr) [inline lib/jekyll/site.rb]: in ruby, you don't include the `()` unless it takes arguments :)

Comment (nickburlett) [inline lib/jekyll/site.rb]: Ah, yes... can you tell it's been a while since I last coded in ruby? (^_^)
I'll fix this and regenerator clear tonight.

**Comment by nickburlett** [comment]
I updated the code to use the clear_cache method instead of manually assigning `@cache = {}` and I've added a test that would fail without the cache clearing in `site.reset`

**Comment by parkr** [comment]
Cool. @alfredxing

**Comment by alfredxing** [comment]
:+1: Thanks! I can't believe I didn't notice this myself earlier...

File: lib/jekyll/site.rb (hunk 2 above)
Comment (parkr) [inline lib/jekyll/site.rb]: should this clear everything, or just the cache?

/cc @alfredxing

Comment (nickburlett) [inline lib/jekyll/site.rb]: I don't think we want to clear the metadata... it should still be a correct representation of what was generated previously. It's just the cache of whether or not a path needs to be regenerated that's now stale.

**Pull request merged**